axes/signals.py

```diff
@@ -1,38 +1,162 @@
+            )
+            log.info('{0} {1}'.format(fail_msg, count_msg))
+    else:
+        # Record failed attempt. Whether or not the IP address or user agent is
+        # used in counting failures is handled elsewhere, so we just record
+        # everything here.
+        AccessAttempt.objects.create(
+            user_agent=user_agent,
+            ip_address=ip_address,
+            username=username,
+            get_data=query2str(request.GET),
+            post_data=query2str(request.POST),
+            http_accept=http_accept,
+            path_info=path_info,
+            failures_since_start=failures,
+        )
+
+        log.info(
+            'AXES: New login failure by {0}. Creating access record.'.format(
+                get_client_str(username, ip_address, user_agent, path_info)
+            )
+        )
+
+    # no matter what, we want to lock them out if they're past the number of
+    # attempts allowed, unless the user is set to notlockable
+    if (
+        failures >= axes_settings.FAILURE_LIMIT and
+        axes_settings.LOCK_OUT_AT_FAILURE and
+        is_user_lockable(request)
+    ):
+        log.warn('AXES: locked out {0} after repeated login attempts.'.format(
+            get_client_str(username, ip_address, user_agent, path_info)
+        ))
+
+        # send signal when someone is locked out.
+        user_locked_out.send(
+            'axes', request=request, username=username, ip_address=ip_address
+        )
+
+
+@receiver(user_logged_in)
+def log_user_logged_in(sender, request, user, **kwargs):
+    """ When a user logs in, update the access log
+    """
+    username = user.get_username()
+    ip_address = get_ip(request)
+    user_agent = request.META.get('HTTP_USER_AGENT', '<unknown>')[:255]
+    path_info = request.META.get('PATH_INFO', '<unknown>')[:255]
+    http_accept = request.META.get('HTTP_ACCEPT', '<unknown>')[:1025]
+    log.info('AXES: Successful login by {0}.'.format(
+        get_client_str(username, ip_address, user_agent, path_info)
+    ))
+
+    if not axes_settings.DISABLE_SUCCESS_ACCESS_LOG:
+        AccessLog.objects.create(
+            user_agent=user_agent,
+            ip_address=ip_address,
+            username=username,
+            http_accept=http_accept,
+            path_info=path_info,
+            trusted=True,
+        )
+
+
+@receiver(user_logged_out)
+def log_user_logged_out(sender, request, user, **kwargs):
+    """ When a user logs out, update the access log
+    """
+    log.info('AXES: Successful logout by {0}.'.format(user))
+
+    if user and not axes_settings.DISABLE_ACCESS_LOG:
+        AccessLog.objects.filter(
             username=user.get_username(),
             logout_time__isnull=True,
-        ).order_by('-attempt_time')[0:1]
-
-        if access_logs:
-            access_log = access_logs[0]
-            access_log.logout_time = now()
-            access_log.save()
+        ).update(logout_time=timezone.now())
```

Comment:
I don't know if it makes any difference but: is it possible to localize the time here?